xenomint/query_sanitizer.go

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2018 The CovenantSQL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenomint
+
+import (
+	"bytes"
+	"database/sql"
+	"strings"
+
+	"github.com/CovenantSQL/CovenantSQL/types"
+	"github.com/CovenantSQL/CovenantSQL/utils/log"
+	"github.com/CovenantSQL/sqlparser"
+	"github.com/pkg/errors"
+)
+
+var (
+	sanitizeFunctionMap = map[string]map[string]bool{
+		"load_extension": nil,
+		"unlikely":       nil,
+		"likelihood":     nil,
+		"likely":         nil,
+		"affinity":       nil,
+		"typeof":         nil,
+		"random":         nil,
+		"randomblob":     nil,
+		"unknown":        nil,
+		"date": {
+			"now":       true,
+			"localtime": true,
+		},
+		"time": {
+			"now":       true,
+			"localtime": true,
+		},
+		"datetime": {
+			"now":       true,
+			"localtime": true,
+		},
+		"julianday": {
+			"now":       true,
+			"localtime": true,
+		},
+		"strftime": {
+			"now":       true,
+			"localtime": true,
+		},
+
+		// all sqlite functions is already ignored, including
+		//"sqlite_offset":             nil,
+		//"sqlite_version":            nil,
+		//"sqlite_source_id":          nil,
+		//"sqlite_log":                nil,
+		//"sqlite_compileoption_used": nil,
+		//"sqlite_rename_table":       nil,
+		//"sqlite_rename_trigger":     nil,
+		//"sqlite_rename_parent":      nil,
+		//"sqlite_record":             nil,
+	}
+)
+
+func convertQueryAndBuildArgs(pattern string, args []types.NamedArg) (containsDDL bool, p string, ifs []interface{}, err error) {
+	var (
+		tokenizer  = sqlparser.NewStringTokenizer(pattern)
+		queryParts []string
+		statements []sqlparser.Statement
+		i          int
+		origQuery  string
+		query      string
+	)
+
+	if queryParts, statements, err = sqlparser.ParseMultiple(tokenizer); err != nil {
+		err = errors.Wrap(err, "parse sql failed")
+		return
+	}
+
+	for i = range queryParts {
+		walkNodes := []sqlparser.SQLNode{statements[i]}
+
+		switch stmt := statements[i].(type) {
+		case *sqlparser.Show:
+			origQuery = queryParts[i]
+
+			switch stmt.Type {
+			case "table":
+				if stmt.ShowCreate {
+					query = "SELECT sql FROM sqlite_master WHERE type = \"table\" AND tbl_name = \"" +
+						stmt.OnTable.Name.String() + "\" AND tbl_name NOT LIKE \"sqlite%\""
```

Comment:
Tooooo many backslashes, why not use back quote?

Reply:
Good idea, working on that.